Add an option for hiscore panel to display the experience of post-99 skills.

geheur opened the pull request:
The plugin now gives a choice between showing 99, showing virtual levels, or showing experience.

![image](https://user-images.githubusercontent.com/41499327/132268909-17f5eac5-c47d-4f24-b3f8-cb44355e0b0c.png)

LlemonDuck commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePlugin.java:
```diff
@@ -69,6 +71,10 @@
 )
 public class HiscorePlugin extends Plugin
 {
+	public static final String CONFIG_GROUP = "hiscore";
+	public static final String CONFIG_KEY_POST_99_DISPLAY = "post99SkillDisplayStyle";
+	public static final String PREVIOUS_CONFIG_KEY_VIRTUAL_LEVEL = "virtualLevels";
```
should be package-local

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePlugin.java:
```diff
@@ -114,6 +125,21 @@ protected void startUp() throws Exception
 		}
 	}
 
+	private void migrateVirtualLevelsDisplayConfig()
```
This should be marked deprecated

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/Post99Display.java:
```diff
@@ -0,0 +1,8 @@
+package net.runelite.client.plugins.hiscore;
```
Needs a license

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscoreConfig.java:
```diff
@@ -55,13 +55,13 @@ default boolean menuOption()
-		name = "Display virtual levels",
-		description = "Display levels over 99 in the hiscore panel"
+		keyName = HiscorePlugin.CONFIG_KEY_POST_99_DISPLAY,
+		name = "Post 99 skills",
```
I don't think this name really makes sense to a user.

geheur commented:
I changed the new config option to read the old config option inside the default value calculation in [4ca81b2](https://github.com/runelite/runelite/pull/14102/commits/4ca81b2c6b8937ec2babe8a428d15baec9f6c2e7). This is a lot simpler but it does mean the old settings will hang around in the config. Not sure which one is better.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java:
```diff
@@ -471,13 +479,29 @@ private void applyHiscoreResult(HiscoreResult result)
+					if ((isSkill && s.getLevel() == 99) || s.getLevel() == 2277)
+					{
+						level = exp;
+						label.setForeground((isSkill && exp == 200_000_000) || exp == 4_600_000_000L
```
`4_600_000_000L` maxed xp will change when sailing comes out

geheur replied:
I updated it to use net.runelite.api.Skill.values().length. This enum is used all over the place and will surely be updated.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java:
```diff
@@ -471,13 +479,29 @@ private void applyHiscoreResult(HiscoreResult result)
-				int level = -1;
-				if (config.virtualLevels() && isSkill && exp > -1L)
+				long level = -1;
+				if (config.maxedSkillsStyle() == MaxedSkillsStyle.VIRTUAL_LEVELS && isSkill && exp > -1L)
```
I believe all of the `exp > -1L` can be replaced with a bool like `hasXp`?

geheur replied:
I simplified a lot of the code here, it seems correct to me.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscoreConfig.java:
```diff
@@ -64,6 +70,17 @@ default boolean virtualLevels()
+	)
+	default MaxedSkillsStyle maxedSkillsStyle()
+	{
+		return virtualLevels() ? VIRTUAL_LEVELS : SHOW_99;
```
Can the this be reverted back into a migrate config?

geheur replied:
Done.

geheur commented:
I addressed raiyni's changes and also changed the format slightly because I think it looks less cluttered:
![image](https://github.com/user-attachments/assets/a4f37595-7d70-4482-8ecf-43d8091d1d4d)

I also removed an unrelated piece of code (if check of player name that is always true).

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java:
```diff
@@ -714,16 +725,20 @@ private void resetEndpoints()
 
 	@VisibleForTesting
-	static String formatLevel(int level)
+	static String formatLevel(long level)
```
this is no longer a level, its either a level or xp. Maybe something like `format(long number)` is better.

geheur replied:
I went with formatNumber(long number).

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/hiscore/HiscorePanel.java:
```diff
@@ -463,40 +468,46 @@ private void applyHiscoreResult(HiscoreResult result)
+							if (s.getLevel() == 99 * skillCount)
+							{
+								level = exp;
+								label.setForeground(exp == 200_000_000L * skillCount ? SKILL_200M_COLOR : SKILL_MAXED_COLOR);
```
use `Experience.MAX_SKILL_XP`

geheur replied:
Done.
